great_expectations/data_context/data_context/abstract_data_context.py

```diff
@@ -995,29 +1029,47 @@ def add_or_update_datasource(
     def add_or_update_datasource(
         self,
         name: str | None = None,
-        datasource: LegacyDatasource | BaseDatasource | None = None,
+        datasource: BaseDatasource | FluentDatasource | LegacyDatasource | None = None,
         **kwargs,
-    ) -> LegacyDatasource | BaseDatasource:
+    ) -> BaseDatasource | FluentDatasource | LegacyDatasource:
         """Add a new Datasource or update an existing one on the context depending on whether
         it already exists or not. The configuration is provided as kwargs.
 
         Args:
             name: The name of the Datasource to add or update.
             datasource: an existing Datasource you wish to persist.
-            kwargs: Any relevant keyword args to use when adding or updating the target Datasource.
+            kwargs: Any relevant keyword args to use when adding or updating the target Datasource named `name`.
 
         Returns:
             The Datasource added or updated by the input `kwargs`.
         """
         self._validate_add_datasource_args(name=name, datasource=datasource)
-        datasource = self._add_datasource(
-            name=name,
-            datasource=datasource,
-            **kwargs,
-        )
-        assert datasource is not None
+        return_datasource: BaseDatasource | FluentDatasource | LegacyDatasource
+        if "type" in kwargs:
+            assert name, 'Fluent Datasource kwargs must include the keyword "name"'
```

Comment:
`non-blocking`
We should capture this in an `overload`. I think What I pasted below should work as is, but LMK if it doesn't.

```python
    @overload
    def add_or_update_datasource(
        self,
        name: str = ...,
        datasource: None = ...,
        type: str = ...,
        **kwargs,
    ) -> FluentDatasource:
       ...

    @overload
    def add_or_update_datasource(
        self,
        name: None = ...,
        datasource: FluentDatasource = ...,
        **kwargs,
    ) -> FluentDatasource:
       ...

    @overload
    def add_or_update_datasource(
        self,
        name: None = ...,
        datasource: BaseDatasource = ...,
        **kwargs,
    ) -> BaseDatasource:
       ...

    @overload
    def add_or_update_datasource(
        self,
        name: None = ...,
        datasource: LegacyDatasource = ...,
        **kwargs,
    ) -> LegacyDatasource:
       ...

    def add_or_update_datasource(
        self,
        name: str | None = None,
        datasource: BaseDatasource | FluentDatasource | LegacyDatasource | None = None,
        **kwargs,
    ) -> BaseDatasource | FluentDatasource | LegacyDatasource:
```